### ISSUE TITLE
Pull chart programmatically with Go API without storing registry credentials

### ISSUE THREAD

**Issue by tomrss** [helm/helm#12144]
Helm version (used as Go dependency): helm.sh/helm/v3 v3.11.3

Hello, I am trying to pull charts in a Go server using the Helm Go API, but I don't want to save registry credentials on filesystem for security reasons.
Is there a way to do it? I didn't succeed finding any registry client configuration or login option that doesn't contemplate storing credentials.

Thanks for the support

**Comment by yxxhero** [comment]
@tomrss could you show your code?

**Comment by tomrss** [comment]
Hi @yxxhero, here is a working example taken from the code I am using:

```go
package main

import (
	"fmt"
	"os"

	"helm.sh/helm/v3/pkg/action"
	"helm.sh/helm/v3/pkg/chart"
	"helm.sh/helm/v3/pkg/chart/loader"
	"helm.sh/helm/v3/pkg/cli"
	"helm.sh/helm/v3/pkg/registry"
)

var credpath string = "./my-creds.json"

func main() {
	// pull chart
	c, err := pullChart("theregistryuser", "theregistrypassword", "thechartregistry", "thechart", "theversion")
	if err != nil {
		panic(err)
	}

	fmt.Printf("Pulled chart: %s\n", c.Name())
	
	// read credential file
	creds, err := os.ReadFile(credpath)
	if err != nil {
		panic(err)
	}

	fmt.Println("Credential file read from filesystem:")
	fmt.Println(string(creds))
}

func logDebug(format string, v ...interface{}){
	fmt.Printf(format, v...)
}

func pullChart(username string, password string, reg string, chart string, version string) (*chart.Chart, error) {
	// registry client
	registryClient, err := registry.NewClient(
		registry.ClientOptDebug(false),
		registry.ClientOptCredentialsFile(credpath),
		registry.ClientOptEnableCache(false),
	)
	if err != nil {
		return nil, err
	}

	// init helm action config
	actionConfig := new(action.Configuration)
	if err := actionConfig.Init(nil, "", "secret", logDebug); err != nil {
		return nil, err
	}

	actionConfig.RegistryClient = registryClient

	// login to registry
	auth := registry.LoginOptBasicAuth(username, password)
	if err := actionConfig.RegistryClient.Login(reg, auth); err != nil {
		return nil, err
	}

	// pull the chart
	pull := action.NewPullWithOpts(action.WithConfig(actionConfig))
	pull.Settings = cli.New() // didn't want to do this but otherwise it goes nil pointer
	pull.ChartPathOptions.Version = version
	pull.DestDir = "."
	if _, err := pull.Run(fmt.Sprintf("oci://%s/%s", reg, chart)); err != nil {
		return nil, err
	}

	// load chart from file
	return loader.LoadFile(fmt.Sprintf("%s-%s.tgz", chart, version))
}

```
The result is that the registry credentials are saved on filesystem

**Comment by tomrss** [comment]
hi @gjenkins8, i think you are right. How about allowing to provide a custom `client.registryAuthorizer` implementation via a `ClientOption`?

**Comment by yxxhero** [comment]
@tomrss PR is welcome.

**Comment by blampe** [comment]
@yxxhero I see this was closed but #12584 is still un-merged. Are there any plans to accept this feature?
